tests/main.rs

```diff
@@ -170,64 +252,213 @@ where
+    let key_pair = PKey::from_rsa(rsa)?;
+
+    let mut x509_name = X509NameBuilder::new()?;
+    x509_name.append_entry_by_text("C", "US")?;
+    x509_name.append_entry_by_text("CN", "localhost")?;
+    let x509_name = x509_name.build();
+
+    let mut cert_builder = X509::builder()?;
+    cert_builder.set_version(2)?;
+    let serial_number = {
+        let mut serial = BigNum::new()?;
+        serial.rand(159, MsbOption::MAYBE_ZERO, false)?;
+        serial.to_asn1_integer()?
+    };
+    cert_builder.set_serial_number(&serial_number)?;
+    cert_builder.set_subject_name(&x509_name)?;
+    cert_builder.set_issuer_name(&x509_name)?;
+    cert_builder.set_pubkey(&key_pair)?;
+    let not_before = Asn1Time::days_from_now(0)?;
+    cert_builder.set_not_before(&not_before)?;
+    let not_after = Asn1Time::days_from_now(365)?;
+    cert_builder.set_not_after(&not_after)?;
+
+    let subject_key_identifier =
+        SubjectKeyIdentifier::new().build(&cert_builder.x509v3_context(None, None))?;
+    cert_builder.append_extension(subject_key_identifier)?;
+
+    cert_builder.sign(&key_pair, MessageDigest::sha256())?;
+    let cert = cert_builder.build();
+
+    Ok((cert, key_pair))
+}
+
 #[test]
 fn it_connects() {
+    let auth_context = Arc::new(Mutex::new(None));
+    let auth_context1 = Arc::clone(&auth_context);
     TestingShim::new(
         |_, _| unreachable!(),
         |_| unreachable!(),
         |_, _, _| unreachable!(),
         |_, _| unreachable!(),
+        move |a| {
+            let mut ac = auth_context1.lock().unwrap();
+            *ac = Some(a.clone());
+            Ok(())
+        },
     )
     .test(|_| {});
+
+    let ac = auth_context.lock().unwrap();
+    assert_eq!(
+        ac.as_ref(),
+        Some(&AuthenticationContext {
+            username: Some(b"username".to_vec()),
+            #[cfg(feature = "tls")]
+            tls_client_certs: None
+        })
+    );
 }
 
 #[test]
 #[cfg(feature = "tls")]
 fn it_connects_tls_server_only() {
     // Client can connect ok without SSL when SSL is enabled on the server.
+    let auth_context = Arc::new(Mutex::new(None));
+    let auth_context1 = Arc::clone(&auth_context);
     TestingShim::new(
         |_, _| unreachable!(),
         |_| unreachable!(),
         |_, _, _| unreachable!(),
         |_, _| unreachable!(),
+        move |a| {
+            let mut ac = auth_context1.lock().unwrap();
+            *ac = Some(a.clone());
+            Ok(())
+        },
     )
-    .with_server_tls()
-    .test(|_| {})
+    .with_tls(false, true, false)
+    .test(|_| {});
+
+    let ac = auth_context.lock().unwrap();
+    assert_eq!(
+        ac.as_ref(),
+        Some(&AuthenticationContext {
+            username: Some(b"username".to_vec()),
+            tls_client_certs: None
+        })
+    );
 }
 
 #[test]
 #[cfg(feature = "tls")]
-fn it_connects_tls_both() {
-    // SSL connection when ssl enabled on server and used by client
+fn it_connects_tls_both_no_client_certs() {
```

Comment:
These tests all look _very_ similar. Could we extract most of the logic to a function with a few arguments that all of these tests call?